### ISSUE TITLE
gccrs cannot generate ET_DYN executables

### ISSUE THREAD

**Issue by CohenArthur** [Rust-GCC/gccrs#556]
I tried this code:

```c
int main(void)
{
    return 0;
}
```

Compiled with `-fPIC`

I expected `gccrs` to produce an executable with an elf type value of `ET_DYN`.

This is what my system's `gcc` does, and this is necessary for compiling executables that are similar to what `rustc` produces.

I have compiled `gccrs` with the C language enabled.

```sh
~/G/cargo-gccrs (check-correct-filetype|✚2…) $ cc -fPIC main.c -o main # cc is a symlink to my system's gcc, /usr/bin/gcc
~/G/cargo-gccrs (check-correct-filetype|✚2…) $ readelf -h main
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              DYN (Shared object file) # Note the e_type value here
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x1020
  Start of program headers:          64 (bytes into file)
  Start of section headers:          14104 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         13
  Size of section headers:           64 (bytes)
  Number of section headers:         28
  Section header string table index: 27
```

Instead, this happened:

```sh
~/G/cargo-gccrs (check-correct-filetype|✚2…) $ gccrs -fPIC main.c -o main
~/G/cargo-gccrs (check-correct-filetype|✚2…) $ readelf -h main
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              EXEC (Executable file) # Note ET_EXEC instead of ET_DYN
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x401140
  Start of program headers:          64 (bytes into file)
  Start of section headers:          47560 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         14
  Size of section headers:           64 (bytes)
  Number of section headers:         41
  Section header string table index: 40
  ```
  
The behavior is similar when trying to create a Position Independant Executable with `-fPIE`.
  
The example provided is using C, but the behavior is similar in Rust. This is simply to get a more accurate comparison.
`rustc` produces executables that are always position independant (Thanks @bjorn3 for [the information](https://github.com/Rust-GCC/cargo-gccrs/issues/35#issuecomment-875371170)), which is a behavior that we might want to reproduce with `gccrs`.

### Meta

According to @philberty's research when debugging me, this has been happening for a while.
My `gccrs` version is `gcc version 11.0.1 20210325 (experimental) (GCC)`

**Comment by bjorn3** [comment]
Does `-fPIE` work? `-fPIC` affects object file generation, while `-fPIE` affects linking I think.

**Comment by CohenArthur** [comment]
> Does `-fPIE` work? `-fPIC` affects object file generation, while `-fPIE` affects linking I think.

Nope, the output is exactly the same. I've tried that but forgot to add it to the issue, I'll do it thanks!

**Comment by dkm** [comment]
-fpic does not imply dynamic, simply code gen yes. You need to use `-shared` or `-static` to force the output result. Maybe `-fPIE` does it

**Comment by dkm** [comment]
```
$ gcc -fPIC -o main main.c
$ file main
main: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=348339ee88b63ec5fd57c7ebcee0ec4667053cc7, for GNU/Linux 3.2.0, not stripped
$ gcc -o main main.c
$ file main
main: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=348339ee88b63ec5fd57c7ebcee0ec4667053cc7, for GNU/Linux 3.2.0, not stripped
$ gcc -o main main.c  -static
$ file main
main: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), statically linked, BuildID[sha1]=abc1157e27c07d6849a0f9f2f574f35f3d0b934a, for GNU/Linux 3.2.0, not stripped
$ gcc -o main main.c  -fpie
$ file main
main: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=348339ee88b63ec5fd57c7ebcee0ec4667053cc7, for GNU/Linux 3.2.0, not stripped

dkm@arrakis:~$ gcc -o main main.c  -fpie
dkm@arrakis:~$ file main
main: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=348339ee88b63ec5fd57c7ebcee0ec4667053cc7, for GNU/Linux 3.2.0, not stripped
$ readelf -h main
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              DYN (Shared object file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x1040
  Start of program headers:          64 (bytes into file)
  Start of section headers:          14600 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         11
  Size of section headers:           64 (bytes)
  Number of section headers:         29
  Section header string table index: 28
$ gcc -o main main.c  -static
$ readelf -h main
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 03 00 00 00 00 00 00 00 00
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - GNU
  ABI Version:                       0
  Type:                              EXEC (Executable file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x401b10
  Start of program headers:          64 (bytes into file)
  Start of section headers:          780784 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         8
  Size of section headers:           64 (bytes)
  Number of section headers:         29
  Section header string table index: 28
```

**Comment by CohenArthur** [comment]
> -fpic does not imply dynamic, simply code gen yes. You need to use `-shared` or `-static` to force the output result. Maybe `-fPIE` does it

I am not trying to generate a shared library, I simply want an executable that is position independant and therefore appears as `ET_DYN`. If I pass `-shared` to my system's gcc, the produced binary does not work correctly and segfaults.

`-static` does not change readelf's output

**Comment by tschwinge** [comment]
Distribution GCC builds are often configured with `--enable-default-pie` (check `gcc -v`).  This makes GCC default to `-fPIE -pie`, resulting in the behavior you reported.  If you override with `-fno-PIE -no-pie`, you should get your `gccrs` configuration's behavior.

---

(I have not yet spent time on understanding what the actual underlying concern/question is.)

**Comment by CohenArthur** [comment]
Fantastic, thanks a lot @tschwinge. I'm assuming `gccrs` distributions will be compiled in the same fashion so it shouldn't be an issue

**Comment by philberty** [comment]
I would suggest trying to compile GCC with that option just to double-check. I am wondering if we need to do something more generic to not rely on this when it comes to cargo-gccrs, I am not sure about the implications of relying on distribution builds to do this or not.

**Comment by dkm** [comment]
Just tried it:
```
> gccrs block_expr1.rs      -O1     -lm  -o ./block_expr1.exe  -fno-pie -fno-PIE -static
> readelf -h block_expr1.exe |grep Type:
  Type:                              EXEC (Executable file)
> gccrs block_expr1.rs     -O1     -lm  -o ./block_expr1.exe  -fno-pie -fno-PIE
> readelf -h block_expr1.exe |grep Type:
  Type:                              DYN (Shared object file)
> gccrs block_expr1.rs     -O1     -lm  -o ./block_expr1.exe
> readelf -h block_expr1.exe |grep Type:
  Type:                              DYN (Shared object file)
```

>  If I pass `-shared` to my system's gcc, the produced binary does not work correctly and segfaults.
 
Sorry, `-shared` will make a library not an executable, hence the segfault, sorry for answering too quickly :/

**Comment by CohenArthur** [comment]
> Sorry, `-shared` will make a library not an executable, hence the segfault, sorry for answering too quickly :/

All good no worries :) Thanks for trying to help me!

**Comment by tschwinge** [comment]
@CohenArthur, you shouldn't rely on what distributors may or may not be doing -- If you reliably want to generate `ET_DYN` executables, then invoke GCC with `-fPIE -pie`.

---

(I still have not yet spent time on understanding what the actual underlying concern/question is.)

**Comment by philberty** [comment]
It seems that rustc creates binaries like this by default:

```
$ rustc test.rs
$ readelf -h test
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              DYN (Shared object file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x5070
  Start of program headers:          64 (bytes into file)
  Start of section headers:          3187136 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         12
  Size of section headers:           64 (bytes)
  Number of section headers:         41
  Section header string table index: 40
```

**Comment by CohenArthur** [comment]
@tschwinge @philberty the idea was not to rely on distributions, no worries. I did not know that in order to create a position independant executable you needed to add `-fPIE` as well as `-pie`. I was originally passing just `-fPIC` and couldn't get the right type, so I asked Philip about it and we looked for it. We thought it might be a bug in the frontend so I opened this issue in the hope that a GCC guru would help :D

I tried compiling the compiler with `--enable-default-pie` and could verify that it works and that it's not a bug in the Rust frontend. I'm gonna be passing `-fPIE -pie` as argument to the compiler for now, but we could decide that this should be the default behavior when compiling rust (or not).
